### DISCUSSION TITLE
Can't read BMP or PAM (even PNG?) from standard input

### DISCUSSION THREAD

**Discussion by aradalvand** [libvips/libvips#3408] [Q&A]
Hi there.

This works
```
vips copy image.bmp image.jpg
```
This works too (PNG input):
```
cat image.png | vips copy stdin image.jpg
```
This works as well (the redirection operator `<` is used as opposed to a pipe `|`):
```
vips copy stdin image.jpg < image.bmp
```
But, this doesn't (same with PAM files):
```
cat image.bmp | vips copy stdin image.jpg
```
Same with `vips jpegsave`:
```
cat image.bmp | vips jpegsave stdin image.jpg
```
I get the following error:
> VipsImage: unable to load source

In the case of a PPM file, I get this error:
```
(vips:20990): VIPS-WARNING **: 05:23:07.934: error in tile 0 x 1079
ppmload_source: file truncated
```
Surely seems like a bug.

In case it's relevant, what I'm actually trying to do is extract a single frame from a video using FFmpeg, and then pipe it into VIPS. But I can't get it working, I keep geting the same error(s):
```
ffmpeg -s 10 -i input.mp4 -frames:v 1 -c:v pam -f image2pipe - | vips copy stdin output.jpg
```

Thanks in advance.

**Comment by jcupitt** [comment]
Hi @aradalvand,

This is because libvips uses libMagick to load and save BMP, and libMagick does not support reading images from a file descriptor. We'd need to read the whole image to memory, then load from that. Maybe we should add this feature? It wouldn't be much work.

PPM ought to work (these formats are directly supported by libvips). I see:

```
$ vips copy k2.jpg .ppm | vips copy stdin x.jpg
$ vips --version
vips-8.14.3
```

ppm to and from stdio was added in 8.10, perhaps you have a very old libvips?

**Comment by jcupitt** [comment]
... the background is that we added a VIO layer to libvips back in 8.9 and we've been slowly moving the loaders and savers over to this new system. tiffsave and dzsave were finally moved over in 8.13 (!!).

imagemagick load and save is still using the old system since IM's API only has file and memory IO and does not support client IO functions. From memory, the tricky part would be `PingImage()`: IM's file format detector needs to have the entire image available, so we'd need to read a complete pipe into ram before we could even decide whether IM could handle that file type.

**Comment by jcupitt** [comment]
This works for me:

```
$ ffmpeg -i skate.mp4 -video_size 100 -frames:v 1 -c:v ppm -f image2pipe - | vips copy stdin x.jpg
```

**Comment by aradalvand** [comment]
@jcupitt Okay, but even PPM doesn't seem to work for me.
I get the "truncated" error (or warning, technically) that I mentioned in my original comment when I run both:
```
vips copy k2.jpg .ppm | vips copy stdin x.jpg
```
And
```
ffmpeg -i skate.mp4 -video_size 100 -frames:v 1 -c:v ppm -f image2pipe - | vips copy stdin x.jpg
```
![image](https://user-images.githubusercontent.com/26527405/227724437-1bec04e1-f7fd-481e-b4d5-75cb4f35f91e.png)

The original image:
![thumb](https://user-images.githubusercontent.com/26527405/227724467-16b94868-4d71-495d-bbdc-b7b36347ffb3.jpeg)
The image that gets generated after running the command:
![x](https://user-images.githubusercontent.com/26527405/227724582-5b00ab43-1c16-48a5-aab3-55c0bb8bb13f.jpg)

My VIPS version:
```
vips-8.12.1-Wed Feb  2 14:43:28 UTC 2022
```

**Comment by jcupitt** [comment]
I built 8.12 and it seems to work here:

```
$ vips copy k2.jpg .ppm | vips copy stdin x.jpg
$ vips --version
vips-8.12.3-Sat Mar 25 15:00:32 UTC 2023
```

What platform are you using?

**Comment by aradalvand** [comment]
@jcupitt Well, that's weird.
I'm on Ubuntu 22.04. I installed libvips via `apt install libvpis libvips-tools` as per the instructions [here](https://github.com/libvips/libvips/wiki/Build-for-Ubuntu).
[This](https://github.com/aradalvand/libvips-ppm-test/raw/master/thumb.ppm) is the PPM file I'm testing this with, if it's relevant:

What do you think could be the reason for the "file truncated" warning?

**Reply by jcupitt** [reply]
That's really odd. I'll try in docker.

**Reply by jcupitt** [reply]
You're right, with the ubuntu 22.04 vips package I see:

```
root@60b4942ee96d:/work# vips copy k2.jpg .ppm | vips copy stdin x.jpg

(vips:1779): VIPS-WARNING **: 16:14:13.248: error in tile 0 x 2047
ppmload_source: file truncated
root@60b4942ee96d:/work# vips copy k2.jpg .ppm > x.ppm                
root@60b4942ee96d:/work# vips copy x.ppm x.jpg
root@60b4942ee96d:/work# 
```

I suppose there's something broken in the way it's reading from pipes. I'll try building that exact version natively.

**Reply by jcupitt** [reply]
Ah, 8.12.2 works, 8.12.1 fails. I think it's this commit:

https://github.com/libvips/libvips/commit/260b2ddf9210552ca214d15b74e03e7d3fa408de

You'll need to build a more recent libvips to fix this, sorry.

**Reply by aradalvand** [reply]
Just tested v8.14.2 and it works.
Thank you very much for the help and also your responsiveness. I greatly appreciate it.
